crates/rattler_cache/src/package_cache/mod.rs

```diff
@@ -98,17 +131,151 @@ impl From<Cancelled> for PackageCacheError {
+            cache_key.sha256.as_ref(),
+            None,
+            None,
+        )
+        .await
+        {
+            Ok(cache_lock) => {
+                cache_entry.last_revision = Some(cache_lock.revision);
+                cache_entry.last_sha256 = cache_lock.sha256;
+                Ok(cache_lock)
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    /// Validate the package, and fetch it if invalid.
+    pub async fn validate_or_fetch<F, Fut, E>(
+        &self,
+        fetch: F,
+        cache_key: &CacheKey,
+        reporter: Option<Arc<dyn CacheReporter>>,
+    ) -> Result<CacheLock, PackageCacheLayerError>
+    where
+        F: (Fn(PathBuf) -> Fut) + Send + 'static,
+        Fut: Future<Output = Result<(), E>> + Send + 'static,
+        E: std::error::Error + Send + Sync + 'static,
+    {
+        let entry = self
+            .packages
+            .entry(cache_key.clone().into())
+            .or_default()
+            .clone();
+
+        let mut cache_entry = entry.lock().await;
+        let cache_path = self.path.join(cache_key.to_string());
+
+        match validate_package_common(
+            cache_path,
+            cache_entry.last_revision,
+            cache_key.sha256.as_ref(),
+            Some(fetch),
+            reporter,
+        )
+        .await
+        {
+            Ok(cache_lock) => {
+                cache_entry.last_revision = Some(cache_lock.revision);
+                cache_entry.last_sha256 = cache_lock.sha256;
+                Ok(cache_lock)
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
 impl PackageCache {
-    /// Constructs a new [`PackageCache`] located at the specified path.
+    /// Constructs a new [`PackageCache`] with only one layer.
     pub fn new(path: impl Into<PathBuf>) -> Self {
-        Self {
-            inner: Arc::new(PackageCacheInner {
+        Self::new_layered(std::iter::once(path.into()))
+    }
+
+    /// Constructs a new [`PackageCache`] located at the specified paths.
+    /// Read-only layers are queried first.
+    /// Within read-only layers, the ordering is defined in this constructor. Ditto for writable layers.
+    pub fn new_layered<I>(paths: I) -> Self
+    where
+        I: IntoIterator,
+        I::Item: Into<PathBuf>,
+    {
+        let layers = paths
+            .into_iter()
+            .map(|path| PackageCacheLayer {
                 path: path.into(),
                 packages: DashMap::default(),
-            }),
+            })
+            .collect();
+
+        Self {
+            inner: Arc::new(PackageCacheInner { layers }),
         }
     }
 
+    /// Returns a tuple containing two sets of layers:
+    /// - A collection of read-only layers.
+    /// - A collection of writable layers.
+    ///
+    /// The permissions are checked at the time of the function call.
+    pub fn split_layers(&self) -> (Vec<&PackageCacheLayer>, Vec<&PackageCacheLayer>) {
+        let readonly_layers = self
```

Comment:
You can use https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition

Reply:
Thank you for the tip ^^ integrated this fix